keps/sig-api-machinery/20190228-priority-and-fairness.md

```diff
@@ -0,0 +1,345 @@
+of the targeted release**.
+
+For enhancements that make changes to code or processes/procedures in core Kubernetes i.e., [kubernetes/kubernetes], we require the following Release Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These checklist items _must_ be updated for the enhancement to be released.
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+This KEP generalizes the existing max-in-flight request handler in the
+apiserver to make more distinctions among requests and provide
+prioritization and fairness among the categories of requests.  An
+outline of the request handling in an apiserver can be found at
+https://speakerdeck.com/sttts/kubernetes-api-codebase-tour?slide=18 .
+
+## Motivation
+
+Today the apiserver has a simple mechanism for protectimg itself
+against CPU and memory overloads: max-in-flight limits for mutating
+and for readonly requests.  Apart from the distinction between
+mutating and readonly, no other distinctions are made among requests;
+consequently there can be undesirable scenarios where one subset of
+the request load crowds out other parts of the request load.  Also,
+the dispatching of these requests against two independent limits is
+not
+[work-conserving](https://en.wikipedia.org/wiki/Work-conserving_scheduler).
+
```

Comment:
Currently my understanding is that max-in-flight puts a ceiling on qps:
https://github.com/kubernetes/apiserver/blob/908ce86a0f4b3f578275e06826d3be48d5096a4c/pkg/server/config.go#L265-L266

Would a goal/side-affect of this enchancement remove that hard limit and is that worth mentioning here?

Reply:
Those are not QPS limits, those are concurrency limits.

Reply:
Ah makes sense, thanks!